### PR TITLE
extending cli to add ability to add a scenario to an existing project

### PR DESCRIPTION
**Pull request by AadilLatif**

### CODE REVIEW

File: tests/test_add_scenario.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        tmpdirname_obj = Path(tmpdirname)
+        shutil.copytree(PYDSS_PROJECT, tmpdirname, dirs_exist_ok=True)
+        settings_path = tmpdirname_obj / "simulation.toml"
+        settings = toml.load(settings_path)
+        settings['Project']['Project Path'] = str(tmpdirname_obj.absolute())
+        toml.dump(settings, open(settings_path, "w"))
+        
+        print(tmpdirname_obj)
+        build_scenario(str(tmpdirname), "test_scenario", str(MAPPING_FILE))
+        os.system("pause")
```
Comment (daniel-thom) [inline tests/test_add_scenario.py:26]: What’s going on here?

Comment (AadilLatif) [reply tests/test_add_scenario.py:26]: oh was doing it for testing I will remove it

File: tests/test_add_scenario.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import shutil
+import toml
+import os
+
+from PyDSS.cli.add_scenario import build_scenario
+
+PYDSS_PROJECT = Path(__file__).parent / "data" / "project"
+MAPPING_FILE = Path(__file__).parent / "data" / "add_scenario" / "controller_map.toml"
+
+def test_add_scenario():
```
Comment (daniel-thom) [inline tests/test_add_scenario.py:13]: If you want to save some logic, you can use this built-in fixture:
```
def test_add_scenario(tmp_path):
```
`tmp_path` is a `Path` directory that will be deleted for you on exit.

File: PyDSS/cli/add_scenario.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    project_path = Path(project_path)
+    controller_mapping_path = Path(controller_mapping)    
+    assert project_path.exists(), "Provided project path does not exist"
+    assert (project_path / "Scenarios").exists(), "provided project is not a valid PyDSS project"
+    assert controller_mapping_path.exists(), "rovided controller_mapping file does not exist"
+    assert controller_mapping_path.suffix.lower() == ".toml", "controller_mapping should be a json file"
+    
+    controller_map = toml.load(controller_mapping_path)
+    mapped_controllers = MappedControllers(**controller_map)
+    print(mapped_controllers)
```
Comment (daniel-thom) [inline PyDSS/cli/add_scenario.py:20]: Does this need to be removed?

File: PyDSS/cli/add_scenario.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            controllers[controller.controller_type] = toml.load(settings_path_obj)
+    scenario_dir = project_path / "Scenarios" / scenerio_name
+    scenario_obj = PyDssScenario(
+            [scenerio_name], list(controllers.keys()), export_modes=None, visualization_types=None
+        )
+    scenario_obj.serialize(str(scenario_dir))
+
+@click.argument("project-path", type=click.Path(exists=True))
+@click.option(
+    "-s", "--scenerio-name",
```
Comment (daniel-thom) [inline PyDSS/cli/add_scenario.py:38]:
```suggestion
    "-s", "--scenario-name",
```

File: PyDSS/cli/add_scenario.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    required=True,
+    help="name of the new scenario",
+)
+@click.option(
+    "-c", "--controller-mapping",
+    required=True,
+    default=None,
+    type=click.Path(exists=True),
+    help="JSON file tha maps controller to contoller defination files",
```
Comment (daniel-thom) [inline PyDSS/cli/add_scenario.py:47]:
```suggestion
    help="JSON file that maps controller to controller definition files",
```

File: PyDSS/cli/add_scenario.py (hunk 3 above)
Comment (AadilLatif) [reply PyDSS/cli/add_scenario.py]: removed

File: PyDSS/cli/add_scenario.py (hunk 5 above)
Comment (AadilLatif) [reply PyDSS/cli/add_scenario.py:46]: fixed

File: PyDSS/cli/add_scenario.py (hunk 4 above)
Comment (AadilLatif) [reply PyDSS/cli/add_scenario.py:37]: fixed

File: tests/test_add_scenario.py (hunk 2 above)
Comment (AadilLatif) [reply tests/test_add_scenario.py:13]: I will just keep it as is

File: PyDSS/cli/add_scenario.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from PyDSS.simulation_input_models import MappedControllers
+from PyDSS.pydss_project import PyDssScenario
+from PyDSS.common import CONTROLLER_TYPES, ControllerType
+
+def build_scenario(project_path:str, scenario_name:str, controller_mapping:str):
+    project_path = Path(project_path)
+    controller_mapping_path = Path(controller_mapping)    
+    assert project_path.exists(), "Provided project path does not exist"
+    assert (project_path / "Scenarios").exists(), "provided project is not a valid PyDSS project"
+    assert controller_mapping_path.exists(), "rovided controller_mapping file does not exist"
```
Comment (daniel-thom) [inline PyDSS/cli/add_scenario.py:15]:
```suggestion
    assert controller_mapping_path.exists(), “provided controller_mapping file does not exist"
```

File: PyDSS/cli/add_scenario.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    assert controller_mapping_path.suffix.lower() == ".toml", "controller_mapping should be a json file"
```
Comment (daniel-thom) [inline PyDSS/cli/add_scenario.py:16]:
```suggestion
    assert controller_mapping_path.suffix.lower() == ".toml", "controller_mapping should be a TOML file"
```

File: PyDSS/cli/add_scenario.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    "-s", "--scenerio-name",
+    required=True,
+    help="name of the new scenario",
+)
+@click.option(
+    "-c", "--controller-mapping",
+    required=True,
+    default=None,
+    type=click.Path(exists=True),
+    help="JSON file that maps controller to contoller defination files",
```
Comment (daniel-thom) [inline PyDSS/cli/add_scenario.py:46]:
```suggestion
    help="JSON file that maps controller type to controller definition files",
```

File: PyDSS/cli/add_scenario.py (hunk 6 above)
Comment (AadilLatif) [reply PyDSS/cli/add_scenario.py:15]: fixed

File: PyDSS/cli/add_scenario.py (hunk 8 above)
Comment (AadilLatif) [reply PyDSS/cli/add_scenario.py:50]: fixed

File: PyDSS/cli/add_scenario.py (hunk 7 above)
Comment (AadilLatif) [reply PyDSS/cli/add_scenario.py:16]: fixxed

**Review by daniel-thom — APPROVED** [review]

**Pull request merged**